Bundle Health / Sanity Check

a10y opened the pull request:
@percyliang I also went ahead and added a rule that will delete the `.sh` files if its corresponding bundle is marked ready or failed. Everything else is as specified in #346

Commit:
#346: Add bundle store health checking functionality

Added a mixin and corresponding implementation to allow health
checking in MultiDiskBundleStore. It goes through and ensures
that the bundle store contains no spurious files, and copies any
spurious data and bundles that are out of sync with the db to a
trash location.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+            3. Any files that don't begin with a UUID string should be moved to trash
+            4. For each bundle marked READY or FAILED, ensure that its dependencies are not located in the bundle
+               directory.
+            5. Bundles that are stored in <UUID>.cid or <UUID>.status/ should be marked as READY or FAILED
```
What do you mean 'stored in <UUID>.cid'?  If the bundle is READY or FAILED, then those files should be moved to trash.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+            4. For each bundle marked READY or FAILED, ensure that its dependencies are not located in the bundle
+               directory.
+            5. Bundles that are stored in <UUID>.cid or <UUID>.status/ should be marked as READY or FAILED
+            6. <UUID>.sh and <UUID>-internal.sh should be gone for bundles marked READY or FAILED
```
Say more accurately, moved to trash.

5 and 6 (and perhaps 4) can be consolidated into: if bundle is done, then make sure it doesn't have random junk associated with it.

percyliang commented:
Another thing we should do: after cleaning up, if the data_hash doesn't exist for a bundle, then compute the data_hash.

Also add a more expensive flag that goes through and makes sure all the data_hashes are accurate.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+            6. <UUID>.sh and <UUID>-internal.sh should be gone for bundles marked READY or FAILED
+        """
+        # Create a trash directory
+        trash_dir = os.path.join(self.codalab_home, 'trash')
```
Let's not move things into a trash directory because this might involve transferring between disks.  Instead, prints out 'rm -r ...' on the things that should be deleted it, and have a force mode that actually executes these commands.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+
+        # Check if bundle is in the database.
+        try:
+            bundle_db = model.get_bundle(uuid)
```
Is this slow with 10K bundles?  If so, we might consider batching these calls...

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+
+
+    def __check_should_trash(self, bundle_path, model):
+        """Checks a path to a bundle to see if it needs to be trashed.
```
bundle_path is misleading - this is actually any path (possibly ending with '.cid')

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+
+        if bundle_db.state in [State.READY, State.FAILED]:
+            # This is the set of paths that should be gone if the bundle is either READY or FAILED
+            # If these paths are still around then something has gone wrong and we should move the bundle to the trash.
```
Shouldn't delete the whole bundle!  Should just delete the dependencies inside the bundle.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,76 @@ def cleanup(self, uuid, dry_run):
+                    return True
+
+            # Bundles marked as READY or FAILED should not end with .cid, .status or .sh
+            if bundle_path.endswith('.cid') or bundle_path.endswith('.status') or bundle_path.endswith('.sh'):
```
If there are other things in the store with <UUID>.<random-ext>, then we should print a warning (but don't delete it just to be safe, since later we'll have .tar.gz).

Commit:
Added -f flag to bs-health-check, warning for files with extensions

Commit:
Separated logic to allow for batch fetching of bundles from DB

a10y commented:
@percyliang If there's a bundle found that has a bad hash, should it delete it, update the data_hash stored in the database or just print a warning?

Commit:
Added checking/computing of data_hash for health check

Now, if a bundle has a NoneType data_hash, the correct hash will
be calculated, and subsequently committed back to the DB if the
--force flag is passed. Also added a new flag, --data-hash
(or -d short form) that will check every bundle to recalculate
data_hash for every bundle and compare against what is in the
database.

Also added --repair flag to fix up the data_hash for bundles
whose data_hash stored in the database does not reflect the
disk contents.

a10y commented:
I just have it print a warning now but that can be changed, all other functionality is present.

a10y commented:
(Fixed a comment)

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,130 @@ def cleanup(self, uuid, dry_run):
+
+            for bundle in bundle_paths:
+                uuid = _get_uuid(bundle)
+                bundle_model = db_bundle_by_uuid[uuid]
```
I get a KeyError when there's a directory with a uuid, but is not in the DB.  This should be treated as to_delete.

a10y replied:
Yeah that was my mistake, I forgot that bundles not in the db will not necessarily have been deleted by this line, only if the -f flag had been used. Should be fixed now.

percyliang commented:
Print the warning for now.  Have a command-line option that fixes the data
hashes.

On Tue, Mar 1, 2016 at 8:04 PM, Andrew notifications@github.com wrote:

> I just have it print a warning now but that can be changed, all other
> functionality is present.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/360#issuecomment-191046405.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,132 @@ def cleanup(self, uuid, dry_run):
+
+        def _check_bundle_paths(bundle_paths, db_bundle_by_uuid):
+            """
+            Takes in a list of bundle paths, and returns a list of paths and subpaths that need to be removed.
```
Document what db_bundle_by_uuid is

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,132 @@ def cleanup(self, uuid, dry_run):
+            for bundle in bundle_paths:
+                uuid = _get_uuid(bundle)
+                bundle_model = db_bundle_by_uuid.get(uuid, None)
+                if bundle_moel == None:
```
typo

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,132 @@ def cleanup(self, uuid, dry_run):
+
+            for bundle in bundle_paths:
+                uuid = _get_uuid(bundle)
+                bundle_model = db_bundle_by_uuid.get(uuid, None)
```
let's just call this bundle instead of bundle_model to be consistent

a10y commented:
Let me integrate that and also fix a few other things, I'll push again soon...

Commit:
Fixed comment

Commit:
Renaming and cleaning up some comments

a10y commented:
Okay, that should wrap it up.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,138 @@ def cleanup(self, uuid, dry_run):
+                            model.update_bundle(bundle, db_update)
+
+
+        print >> sys.stderr, 'Deleted %d objects from the bundle store' % trash_count
```
Only print this if --force; print different message otherwise

a10y replied:
I felt like having these summary statistics might be nice as well for the dry-run, but I can also hide them

percyliang replied:
Summary statistics are good - just don't say 'Deleted' when they haven't
been deleted.  Print 'use --force to actually delete'.

On Wed, Mar 2, 2016 at 12:06 AM, Andrew notifications@github.com wrote:

> In codalab/lib/bundle_store.py
> https://github.com/codalab/codalab-cli/pull/360#discussion_r54690493:
> 
> > -                    data_hash = '0x%s' % path_util.hash_directory(bundle_path, dirs_and_files)
> > -                    if bundle.data_hash == None:
> > -                        data_hash_recomputed += 1
> > -                        print >> sys.stderr, 'Giving bundle %s data_hash %s' % (bundle_path, data_hash)
> > -                        if force:
> > -                            db_update = dict(data_hash=data_hash)
> > -                            model.update_bundle(bundle, db_update)
> > -                    elif compute_data_hash and data_hash != bundle.data_hash:
> > -                        data_hash_recomputed += 1
> > -                        print >> sys.stderr, 'Bundle %s should have data_hash %s, actual digest is %s' % (bundle_path, bundle.data_hash, data_hash)
> > -                        if repair_hashes and force:
> > -                            db_update = dict(data_hash=data_hash)
> > -                            model.update_bundle(bundle, db_update)
> >   +
> >   +
> > -        print >> sys.stderr, 'Deleted %d objects from the bundle store' % trash_count
> 
> I felt like having these summary statistics might be nice as well for the
> dry-run, but I can also hide them
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codalab/codalab-cli/pull/360/files#r54690493.

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,138 @@ def cleanup(self, uuid, dry_run):
+        |compute_data_hash|: If True, compute the data_hash for every single bundle ourselves and see if it's consistent with what's in
+                             the database. False by default.
+        """
+        UUID_REGEX = re.compile(r'^(0x[0-9a-z]{32})')
```
Can't we use the one in spec_util?

a10y replied:
UUID_REGEX in spec_util assumes that the entire string is the UUID, here we just wanted to do regex matching to see if the filename begins with a full uuid. I can use spec_util.UUID_STR to build this regex though

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,138 @@ def cleanup(self, uuid, dry_run):
+        UUID_REGEX = re.compile(r'^(0x[0-9a-z]{32})')
+
+        def _delete_path(loc):
+            cmd = 'rm -r %s' % loc
```
Quote this just in case

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,138 @@ def cleanup(self, uuid, dry_run):
+        trash_count = 0
+
+        for partition in partitions:
+            partition_path = os.path.join(self.partitions, partition, MultiDiskBundleStore.DATA_SUBDIRECTORY)
```
Print out 'Looking for trash in partition ...'

percyliang commented on codalab/lib/bundle_store.py:
```diff
@@ -383,3 +395,138 @@ def cleanup(self, uuid, dry_run):
+            # Check for each bundle if we need to compute its data_hash
+            data_hash_recomputed = 0
+
+            for bundle_path in bundle_paths:
```
Print out 'checking data_hash....'

Commit:
Cosmetic changes

Commit:
Change how statistics are displayed based on --force flag

a10y commented:
@percyliang LGTY?

percyliang commented:
LGTM
